src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+    thread_state original = state.state;
+    state.transition_to(new_state);
+    log_transition(state.thread_id, state.task_id, original, new_state, message);
+  }
+
+  /**
+   * throw a java exception using the cached jvm/env.
+   */
   void throw_java_exception(const char* ex_class_name, const char* msg) {
     JNIEnv *env = cudf::jni::get_jni_env(jvm);
     cudf::jni::throw_java_exception(env, ex_class_name, msg);
   }
 
-  void *do_allocate(std::size_t num_bytes, rmm::cuda_stream_view stream) override {
-    auto tid = static_cast<long>(pthread_self());
-    {
-      std::scoped_lock lock(state_mutex);
-      // pre allocate checks
-      auto thread = threads.find(tid);
+  /**
+   * This is a watchdog to prevent us from live locking. It should be called before we throw an RetryOOM
+   * or a SplitAndRetryOOM to know if we actually should throw something else. 
+   */
+  void check_before_oom(full_thread_state & state, std::unique_lock<std::mutex> & lock) {
+    // The limit is an arbitrary number, large enough that we should not hit it in "normal"
+    // operation, but also small enough that we can detect a livelock fairly quickly.
+    // In testing it looks like it is a few ms if in a tight loop, not including spill
+    // overhead
+    if (state.num_times_retried + 1 > 500) { 
+      throw_java_exception(JAVA_OOM_CLASS, "retry limit exceeded");
+    }
+    state.num_times_retried++;
+  }
+
+  void throw_retry_oom(const char* msg, full_thread_state & state, std::unique_lock<std::mutex> & lock) {
+    check_before_oom(state, lock);
+    throw_java_exception(RETRY_OOM_CLASS, "task should retry operation");
+  }
+
+  void throw_split_n_retry_oom(const char* msg, full_thread_state & state, std::unique_lock<std::mutex> & lock) {
+    check_before_oom(state, lock);
+    throw_java_exception(SPLIT_AND_RETRY_OOM_CLASS, "task should split input and retry operation");
+  }
+
+  /**
+   * Internal implementation that will block a thread until it is ready to continue.
+   */
+  void block_thread_until_ready(std::unique_lock<std::mutex> & lock) {
+    bool done = false;
+    bool first_time = true;
+    auto thread_id = static_cast<long>(pthread_self());
+    // Because this is called from alloc as well as from the public faceing block_thread_until_ready
+    // there are states that should only show up in relation to alloc failing. These include
+    // TASK_BUFN_THROW and TASK_SPLIT_THROW. They should never happen unless this is being called
+    // from within an alloc.
+    while (!done) {
+      auto thread = threads.find(thread_id);
+      if (thread != threads.end()) {
+        switch (thread->second.state) {
+        case TASK_BLOCKED:
+        // fall through
+        case TASK_BUFN:
+        // fall through
+        case SHUFFLE_BLOCKED:
+          log_status("WAITING", thread_id, thread->second.task_id, thread->second.state);
+          task_wake_condition.wait(lock);
+          task_has_woken_condition.notify_all();
```

Comment:
This doesn't seem to account for spurious wakeups.  Seems we need to check the state before notifying others that this thread is waking up.